### ISSUE TITLE
Legion: shardrefine slow startup on Frontier

### ISSUE THREAD

**Issue by syamajala** [StanfordLegion/legion#1581]
@elliottslaughter pushed a fix to allow us to run a rank/node configuration on Frontier. I'm not seeing the OOM problem that I was seeing when I run rank/gpu with shardrefine, so I did some scaling runs. 

I am seeing a slow startup as we've seen in the past: http://sapling2.stanford.edu/~seshu/s3d_subrank/shardrefine/startup_legion_prof.html

Some profiles are available here: http://sapling2.stanford.edu/~seshu/s3d_subrank/shardrefine/

**Comment by syamajala** [comment]
I also got some stacktraces from a 128 node run at two points in time if those are useful: http://sapling2.stanford.edu/~seshu/s3d_subrank/shardrefine/stacktraces/

**Comment by lightsighter** [comment]
You're sure you've enabled functional projection functors?

There are two options for debugging this:

1. Either you make a small example for me to play with on sapling
2. I'm going to give you some changes that are going to dump a metric ton of logging to help us understand what is going on.

**Comment by lightsighter** [comment]
Also, the line numbers are out-of-date with respect to the current `shardrefine` branch which makes it hard to understand what is going wrong. What hash was this built with?

**Comment by syamajala** [comment]
From what I remember @elliottslaughter merged the changes for function projection functors. He will have to confirm that though. 

For debugging this why dont we do both options? Given the issues we have with sapling I may not be able to reproduce this there, so you should try to add extra logging as well in the meantime while I see what I can do on sapling. Alex pinged action@cs for an update and they did not have any news other than they will continue looking at sapling. 

From what I see the branch that @elliottslaughter made was rebased on to this commit of shardrefine:
```
commit e264eb0bd5fe7f66c596796a251dd925beec022b (origin/shardrefine)
Merge: 0ba740529 e881d2dda
Author: Mike <mebauer@cs.stanford.edu>
Date:   Tue Oct 31 00:58:17 2023 -0700

    Merge branch 'control_replication' into shardrefine
```

It is the regent-fix-hip-multigpu-shardrefine branch.

**Comment by elliottslaughter** [comment]
> From what I remember @elliottslaughter merged the changes for functional projection functors. He will have to confirm that though.

Yes, it merged into `shardrefine` a long time ago.

> From what I see the branch that @elliottslaughter made was rebased on to this commit of shardrefine: ... the regent-fix-hip-multigpu-shardrefine branch.

`regent-fix-hip-multigpu` (without `-shardrefine`) has merged to `master`, so once we merge `master` -> `control_replication` -> `shardrefine` this will be available as well and we can prune the extra branches.

**Comment by syamajala** [comment]
I am not able to run S3D on sapling right now it keeps hanging at startup with processes in D state.

**Comment by lightsighter** [comment]
Can you try with the most recent `shardrefine` branch at scale. I checked in a slightly more optimized data structure for the projection tree exchange. I want to see if it is blowing up just because STL is bad or because there is an asymptotic overhead.

**Comment by syamajala** [comment]
Its looking much better: http://sapling2.stanford.edu/~seshu/s3d_subrank/shardrefine/startup_legion_prof.html

Down to 15 seconds from 500 seconds at 64 nodes. I submitted up to 2048 nodes so we'll see what happens with the rest.

**Comment by lightsighter** [comment]
Please get some profiles for these runs as well so we can see how things are changing.

**Comment by syamajala** [comment]
Here is the startup time up to 2048 nodes: http://sapling.stanford.edu/~seshu/s3d_subrank/startup_legion_prof.html

Iteration time: http://sapling.stanford.edu/~seshu/s3d_subrank/average_21to30_iteration.html

legion_prof --archive profiles are here: http://sapling.stanford.edu/~seshu/s3d_subrank/

View them like this: https://legion.stanford.edu/prof-viewer/?url=https://sapling2.stanford.edu/~seshu/s3d_subrank/pwave_x_1024_hept/legion_prof/

Still waiting on 4096 and 8192 node runs.

**Comment by lightsighter** [comment]
Did you get backtrace samples from the slow start-up at 8K nodes?

**Comment by syamajala** [comment]
The 8192 run actually worked and completed. We have scaled to the full machine on Frontier. 

The 4096 run did not finish starting up. I tried to get some stack traces from that. They are here: http://sapling2.stanford.edu/~seshu/s3d_subrank/startup/

**Comment by lightsighter** [comment]
The backtraces from 4K are not interesting. Many of the files are actually empty suggesting that those processes never even started.

Did you get backtraces for 8K nodes?

**Comment by syamajala** [comment]
The 4096 node run completed when I resubmitted. Here are the final results:

Startup: http://sapling2.stanford.edu/~seshu/s3d_subrank/startup_legion_prof.html

Iteration: http://sapling2.stanford.edu/~seshu/s3d_subrank/average_21to30_iteration.html

**Comment by lightsighter** [comment]
I'm not satisfied. I would still like to understand why the start-up is not scalable. Please get some backtraces in the middle of the startup for either the 2K, 4K, or 8K runs. There are enough nodes there that just sampling each node should give us enough of a clue as to what is going on.

**Comment by syamajala** [comment]
Here is a test run I did at 128 nodes. I set it up to dump stack traces with gdb every 15 seconds.

http://sapling2.stanford.edu/~seshu/s3d_subrank/startup/

2048 is in the queue.

**Comment by syamajala** [comment]
Im not sure if my script is working reliably. A lot of the stack trace files seem to be empty and I see gdb ending up in a D state. Will have to try something different.

**Comment by syamajala** [comment]
He haven't been able to get stack trace dumping to work reliably. I might try to profile with HPCToolkit or gprof.

**Comment by syamajala** [comment]
This worked on Perlmutter as well. Scaled to 1536 nodes full machine. Have one more machine I am planning on scaling on. Hopefully I will finish that in the next day or two.

**Comment by lightsighter** [comment]
I commented out what I think is the slow part of the remaining start-up. Can we generate results with the latest shardrefine branch on one of the machines?

**Comment by syamajala** [comment]
I did a few smaller test runs first to make sure the build works and I'm seeing a lot of these messages: `[22 - 7fffcbc0df80]    0.005833 {5}{mailbox}: Failed to send message: 111` it still seems to run to completion though.

**Comment by syamajala** [comment]
The 2048 and 4096 node runs did not startup. The 1024 did but I did not see much improvement in startup time. It was about 55 seconds the old build was 59 seconds.

**Comment by lightsighter** [comment]
> The 2048 and 4096 node runs did not startup. 

At what point did they stop running? There are some known issues in Realm bootstrapping right now.

>  The 1024 did but I did not see much improvement in startup time. It was about 55 seconds the old build was 59 seconds.

I'd still like to see some backtrace samples to see where the time is going.

**Comment by syamajala** [comment]
I am actually able to get profiles with hpctoolkit at small scales. Im going to go back to the commit of shardrefine that worked and try profiling with hpctoolkit at larger scales.

**Comment by syamajala** [comment]
Its been a little hit or miss. Im still waiting for the larger runs, but I got a profile from 512 nodes and 4 nodes. Any hints for what I'm looking for?

**Comment by lightsighter** [comment]
I'm just looking for backtraces that suggest where the time is being spent in the start-up. If it was like before it is somewhere in the logical analysis, but hard to say where that is without backtraces. A few hundred different backtraces during start-up should give us a pretty good sampling of where the time is being spent.

**Comment by syamajala** [comment]
Here is the stuff Im seeing in legion_analysis.cc:

![2023-11-13-192157_2551x1280_scrot](https://github.com/StanfordLegion/legion/assets/45757/795115dc-69d3-4e9f-9db1-52bcd3f09311)

**Comment by lightsighter** [comment]
That doesn't really make sense. If I'm interpreting the percentages on the right hand side correctly, 60% of the time is being spent somewhere else, and that is the 60% we're interested in. This is all just trace capture stuff and we know that is scalable. 

I think just sampling backtraces uniformly across all the shards will be more valuable. Pick a stopping time. Capture backtraces from all shards. Resume for a few seconds and then do it again. Do that three or four times. That should give us some representative data.

**Comment by lightsighter** [comment]
Also check to make sure your profiler here understands Realm's user-level threading. If it doesn't a lot of these results are going to be quite confusing. That's why backtraces with `-ll:force_kthreads` are probably going to be more meaningful.

**Comment by syamajala** [comment]
I will try dumping stacktraces again. The script I made before wasnt working reliably at larger scales, but I'll try simplifying it a bit.

**Comment by lightsighter** [comment]
Does the debugger that you're using not having something equivalent to `thread apply all bt`?

**Comment by syamajala** [comment]
I am using gdb, but on some nodes its getting stuck in a D state when it attaches.

**Comment by lightsighter** [comment]
I would rate limit my script to only attach on a few nodes at a time. You might be overwhelming (a poorly implemented) filesystem.

**Comment by syamajala** [comment]
I tried using the nvme drives on the compute nodes so each node is just writing 1 file to a local disk. Here is the launch.sh script:
```
#!/bin/bash
sleep_time=( [1]=3 [2]=3 [4]=3 [8]=3 [16]=6 [32]=6 [64]=6 [128]=10 [256]=10 [512]=20 [1024]=30 [2048]=50 [4096]=110 [8192]=300 )
$@ &
S3D_PID=$!
sleep ${sleep_time[$SLURM_JOB_NUM_NODES]}
outdir=/mnt/bb/seshuy
gdb -p $S3D_PID -batch -quiet -ex "thread apply all bt" 2>&1 > "$outdir/bt_$(hostname)_$S3D_PID.log"
wait $S3D_PID
```

It should be `srun -n 1024 --ntasks-per-node 1 ... launch.sh s3d.x` so you start s3d, sleep for however many seconds and then start gdb, and there should only be 1 gdb process/node writing a single file to its local nvme drive that will get copied over to the network drive after the s3d job step has completed, but when I looked at some of the nodes in the 1024 node job, gdb is still in a D state and the log on /mnt/bb/seshuy/ is empty.

**Comment by lightsighter** [comment]
I suspect the problem is with gdb probing at the whole binary to load all the symbols when it starts up and the file system becoming unhappy...

**Comment by syamajala** [comment]
I made a version that uses sbcast to copy the shared libraries and executable to the nvme drives on the compute nodes: https://docs.olcf.ornl.gov/systems/frontier_user_guide.html#sbcast-your-executable-and-libraries

I was seeing a crash when attaching gdb at startup in Realm when it tries to create shmem, it would work when I didnt attach gdb though, but there were messages about failing to resize a shmem region. I know a bunch of stuff has been changing in Realm, so I upgraded to the latest shardrefine and it works up to 8 nodes, but at 16 nodes I'm seeing this now: 
```
[0 - 7ffe43ffeb40]    7.041922 {5}{runtime}: [error 93] LEGION ERROR: Unable to find entry for color 9 in index space 20. (from file /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/legion/region_tree.cc:8102)
For more information see:
http://legion.stanford.edu/messages/error_code.html#error_code_93
```

Here is a stack trace:
```
[15] Thread 6 (Thread 0x7ffe41ff4b40 (LWP 33715) "s3d.x"):                                                             
[15] #0  0x00007fffe4af274f in wait4 () from /lib64/libc.so.6
[15] #1  0x00007fffe4a69ba7 in do_system () from /lib64/libc.so.6
[15] #2  0x00007fffdedf7776 in gasneti_system_redirected () from /mnt/bb/seshuy/librealm.so.1
[15] #3  0x00007fffdedf711b in gasneti_bt_gdb () from /mnt/bb/seshuy/librealm.so.1
[15] #4  0x00007fffdeded94f in gasneti_print_backtrace () from /mnt/bb/seshuy/librealm.so.1
[15] #5  0x00007fffdef0adaa in gasneti_defaultSignalHandler () from /mnt/bb/seshuy/librealm.so.1
[15] #6  <signal handler called>                                                                                       
[15] #7  0x00007fffe4a5bcbb in raise () from /lib64/libc.so.6                          
[15] #8  0x00007fffe4a5d355 in abort () from /lib64/libc.so.6                          
[15] #9  0x00007fffe1beccdf in Legion::Internal::Runtime::report_error_message (id=93, file_name=0x7fffe0bbb2cf "/lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/legion/region_tree.cc", line=8129, message=0x7ffe41f9973
0 "Unable to find entry for color 9 in index space 20.") at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/legion/runtime.cc:32159
[15] #10 0x00007fffe1b39c23 in Legion::Internal::IndexSpaceNode::get_child (this=0x7ff944159a00, c=9, defer=0x0, can_fail=false) at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/legion/region_tree.cc:8127
[15] #11 0x00007fffe1b3aa5c in Legion::Internal::RegionTreeForest::get_index_partition (this=0x688d780, parent=..., color=9) at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/legion/region_tree.cc:922
[15] #12 0x00007fffe1c54a4a in Legion::Internal::Runtime::get_index_partition (this=0x7600ea0, parent=..., color=9) at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/legion/runtime.cc:18315
[15] #13 0x00007fffe1734489 in Legion::Runtime::get_index_partition (this=0x688d740, parent=..., color=9) at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/legion/legion.cc:5255
[15] #14 0x00007fffe2c7bda4 in legion_terra_index_cross_product_get_subpartition_by_color_domain_point (runtime_=..., prod=..., color_=...) at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/bindings/regent/regent_partitions.
cc:645                                                                                                                                                                                                                                         
[15] #15 0x00007fffe49c5531 in $<main> () from /mnt/bb/seshuy/libregent_tasks.so                                                                                                                                                               
[15] #16 0x00007fffe4910bcd in $__regent_task_main_primary () from /mnt/bb/seshuy/libregent_tasks.so                                                                                                                                           
[15] #17 0x00007fffde1a9c3b in Realm::LocalTaskProcessor::execute_task (this=0x67ef980, func_id=2839, task_args=...) at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/realm/proc_impl.cc:1175
[15] #18 0x00007fffde231fc0 in Realm::Task::execute_on_processor (this=0x7ff8000ca040, p=...) at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/realm/tasks.cc:326 
[15] #19 0x00007fffde23640c in Realm::KernelThreadTaskScheduler::execute_task (this=0x6694c60, task=0x7ff8000ca040) at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/realm/tasks.cc:1421
[15] #20 0x00007fffde2351a0 in Realm::ThreadedTaskScheduler::scheduler_loop (this=0x6694c60) at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/realm/tasks.cc:1160
[15] #21 0x00007fffde23577a in Realm::ThreadedTaskScheduler::scheduler_loop_wlock (this=0x6694c60) at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/realm/tasks.cc:1272
[15] #22 0x00007fffde24723c in Realm::Thread::thread_entry_wrapper<Realm::ThreadedTaskScheduler, &Realm::ThreadedTaskScheduler::scheduler_loop_wlock> (obj=0x6694c60) at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/
realm/threads.inl:97                                                                                                   
[15] #23 0x00007fffde24a47a in Realm::KernelThread::pthread_entry (data=0x688c570) at /lustre/orion/cmb138/scratch/seshuy/legion_s3d_subrank/legion/runtime/realm/threads.cc:781
[15] #24 0x00007fffe88936ea in start_thread () from /lib64/libpthread.so.0
[15] #25 0x00007fffe4b28a6f in clone () from /lib64/libc.so.6
```

I am on this commit of shardrefine:
```
commit cd49b6b5a153d02de2244fc4666e4da587ed98bc (HEAD -> shardrefine, origin/shardrefine)
Author: Mike <mebauer@cs.stanford.edu>
Date:   Wed Nov 15 02:04:58 2023 -0800

    legion: adding an optimization for using congruent equivalence sets across fields when making new equivalence sets
```

**Comment by syamajala** [comment]
I found that shmem crash with the old build depends very heavily on when we attach with gdb, just changing the sleep time seems to make it go away. Another thing I found is that gdb does eventually get out of the D state but by the time it does s3d will have already exited. Maybe I might try not sleeping at all at larger scales and see what happens.

**Comment by syamajala** [comment]
I guess its possible to write the symbol table to a file and have gdb reuse that file when it runs a program. I'm not sure if that will work with attach, but I'll give it a try and see. Or maybe we can use LAUNCHER='gdb --args'

**Comment by lightsighter** [comment]
> I am on this commit of shardrefine:

You should check the CI before switching to new commits of a development branch:
https://gitlab.com/StanfordLegion/legion/-/pipelines/1073205477

**Comment by lightsighter** [comment]
Closing this now that shardrefine is merged.
